components/suggest/src/error.rs

```diff
@@ -29,15 +32,35 @@ pub(crate) enum Error {
 #[derive(Debug, thiserror::Error)]
 #[non_exhaustive]
 pub enum SuggestApiError {
+    #[error("Network error: {reason}")]
+    NetworkError { reason: String },
+    // The application interrupted a request
+    #[error("Interrupted")]
+    Interrupted,
     #[error("Other error: {reason}")]
     Other { reason: String },
 }
 
-impl From<Error> for SuggestApiError {
-    /// Converts an internal component error to a public application error.
-    fn from(error: Error) -> Self {
-        Self::Other {
-            reason: error.to_string(),
+// Define how our internal errors are handled and converted to external errors
+// See `support/error/README.md` for how this works, especially the warning about PII.
+impl GetErrorHandling for Error {
+    type ExternalError = SuggestApiError;
+
+    fn get_error_handling(&self) -> ErrorHandling<Self::ExternalError> {
+        match self {
+            // Do nothing for interrupted errors, this is just normal operation.
+            Self::Interrupted(_) => ErrorHandling::convert(SuggestApiError::Interrupted),
+            // Network errors are fairly common so let's log, but not report them.
+            Self::RemoteSettings(RemoteSettingsError::RequestError(_)) => {
```

Comment:
From what I could tell from Sentry, this was the overwhelming source of network errors.  However, are there other potential sources?

Reply:
I wonder if we should also treat `ResponseError` (and maybe `BackoffError`, too?) as a network error...a 503 or a bad response from Remote Settings could be interesting to note, but doesn't feel super actionable as far as there being something for us to fix in the client. WDYT?

Reply:
I think that last part is the most important: what actions should the client take?  With that in mind I updated the handling:
- `SuggestApiError::Network` now only happens when we see `viaduct::NetworkError`.  If a client sees a network error, they should retry when the network comes back online.  I targetted this very specifically at `viaduct::NetworkError`, because AFAIK, there's no reason to believe that the other errors are transient in the same way -- but please correct me if I'm wrong about this.
- I added `SuggestApiError::Backoff` which corresponds to `remove_settings::BackoffError`, which implies a very specific action for the client.
- All other networky errors get mapped to `SuggestApiError::Other`.  My reasoning is the client should just treat these as something bad happened and try to recover as best as they can, without any expectation that sending the same request again will work.

What do you think about that?  This is mostly based on my gut feelings, so I'm happy to take another pass and update the logic again.

Reply:
That sounds good to me!

> there's no reason to believe that the other errors are transient in the same way

We've seen Remote Settings return 503s occasionally, so the client could retry those, too—but let's not worry about them now; Viaduct network errors are so noisy by comparison. We can always revisit reporting these HTTP errors if they show up frequently in Sentry.